syntax = "proto3";

package proto;


// Comando para gerar o proto:
// python -m grpc_tools.protoc --proto_path=. .\proto\message.proto --python_out=. --grpc_python_out=.


// Como escrever em PROTO
// https://developers.google.com/protocol-buffers/docs/proto3


// Como enviar um array de dados
// https://stackoverflow.com/questions/43167762/how-to-return-an-array-in-protobuf-service-rpc
service Message{

 // Chamar a função Login do servidor passando a mensagem LOGIN do cliente
 rpc Login(MessageClient) returns (LoginResponse) {}
 // Chama a função do servidor para cadastrar um novo usuário
 rpc Create(MessageClient) returns (Response) {}
 // Chama a função do servidor para visualizar o álbum
 rpc Album(MessageClient) returns (AlbumResponse) {}
 // Chama a função do servidor que realiza a compra e exibe as figurinhas
 rpc Buy(MessageClient) returns (AlbumResponse) {}
 // Chama a função venda e mostra as figuras adquiridas
 rpc Sell(MessageClient) returns (SellResponse) {}
  // Chama a função de criação de trocas e response se criou ou não
 rpc CreateTrade(MessageClient) returns (Response) {}
   // Chama a função para exibir todas as trocas cadastradas
 rpc ListTrade(MessageClient) returns (ListTradeResponse) {}
    // Chama a função para exibir todas as trocas cadastradas
 rpc Trade(MessageClient) returns (Response) {}

}

message MessageClient{
 optional int32 idUser = 1;
 optional int32 idFigure = 2;
 optional int32 offer = 3;
 optional int32 taking = 4;
 optional string name = 5;
 optional string password = 6;
 optional int32 idTrade = 7;
}

message Figure{
 int32 idFigure = 1;
 string name = 2;
 string rarity = 3;
 optional string path = 4;
 optional int32 quantity = 5;
 optional int32 idUser = 6;
}

message Trade{
 string name = 1;
 int32  idTrade = 2;
 int32 offerID = 3;
 string offerName = 4;
 string offerRarity = 5;
 int32 takingID = 6;
 string takingName = 7;
 string takingRarity = 8;
}

message ListTradeResponse{
 bool response = 1;
 repeated Trade list = 2;
}

message User{
 int32 idUser = 1;
 string name = 2;
 float balance = 3;
 string password = 4;
 bool showcard = 5;
}

message LoginResponse{
 bool response = 1;
 User user = 2;
 optional Figure figure = 3;
}

message AlbumResponse{
 bool response = 1;
 optional bool complete = 2;
 optional int32 idUser = 3;
 optional Figure special = 4;
 repeated Figure figures = 5;
 optional float balance = 6;
}

message SellResponse{
 bool response = 1;
 float price = 2;
 string name = 3;
}

message Response{
 bool response = 1;
}